dataset/src/main/scala/frameless/macros/ColumnMacros.scala

```diff
@@ -0,0 +1,377 @@
+package frameless.macros
+
+import frameless.{TypedColumn, TypedEncoder}
+import org.apache.spark.sql.ColumnName
+
+import scala.collection.immutable.Queue
+import scala.reflect.macros.{TypecheckException, whitebox}
+
+class ColumnMacros(val c: whitebox.Context) {
+  import c.universe._
+
+  private val TypedExpressionEncoder = reify(frameless.TypedExpressionEncoder)
+  private val TypedDataset = reify(frameless.TypedDataset)
+  private val ColumnName = weakTypeOf[ColumnName]
+
+  private def toColumn[A : WeakTypeTag, B : WeakTypeTag](
+    selectorStr: String,
+    encoder: c.Expr[TypedEncoder[B]]
+  ): Tree = {
+
+    val A = weakTypeOf[A].dealias
+    val B = weakTypeOf[B].dealias
+
+    val typedCol = appliedType(
+      weakTypeOf[TypedColumn[_, _]].typeConstructor, A, B
+    )
+
+
+    val datasetCol = c.typecheck(
+      q"new $ColumnName($selectorStr).as[$B]($TypedExpressionEncoder.apply[$B]($encoder))"
```

Comment:
I changed this to use `ColumnName` so that it doesn't refer to `c.prefix` - that means it's friendly to chaining like

```scala
ds.groupBy(_.a).agg(sum(ds(_.b)), count(ds(_.b)).selectExpr(x => (x._1, x._2 / x._3))
```

And the `ds` references in `agg` won't hurt anything because they are free of `ds` itself. The downside is you could use a column from a different dataset and it would compile (and even work if the other dataset has a same-named column).